File: Medium/176_RemoveAllAdjacentDuplicatesInStringII.java
class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Character> st= new Stack<Character>();
        int count=1;char curr=s.charAt(0);
        st.push(curr);
        for(int i=1;i<s.length();i++){
            //System.out.println(count+" "+curr);
            if(count==0){count++;curr=s.charAt(i);st.push(curr);}
            else if(curr==s.charAt(i)){
                count++;
                if(count==k){
                    while(count!=1){
                        st.pop();count--;
                    }
                    count=0;
                    if(!st.isEmpty()){
                        curr=st.pop();
                        count=1;
                        while(!st.isEmpty() && curr==st.peek()){
                            count++;st.pop();
                        }
                        int c=count;
                        while(c!=0){c--;st.push(curr);}
                    }
                    
                }
                else{
                    st.push(curr);
                }
            }
            else{
                count=1;curr=s.charAt(i);
                st.push(curr);
            }
        }
        StringBuilder sb=new StringBuilder();
        while(!st.isEmpty()){
            sb.append(st.pop());
        }
        return sb.reverse().toString();
    }
}

File: Medium/176_RemoveAllAdjacentDuplicatesInStringII.py
class Node():
    val = None
    count = None
    def __init__(self, val, count):
        self.val = val
        self.count = count
    
    def __repr__(self):
        return "{},{}".format(self.val, self.count)
    
class Solution(object):
    def removeDuplicates(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        node_stack = []
        for character in s:
            node_top = node_stack[-1] if node_stack else None
            if node_top and node_top.val == character:
                curr_count = node_top.count + 1
                if curr_count == k:
                    while node_top and node_top.val == character:
                        node_stack.pop()
                        node_top = node_stack[-1] if node_stack else None
                else:
                    node_stack.append(Node(character, curr_count))
            else:
                node_stack.append(Node(character, 1))
        new_string = ""
        for node in node_stack:
            new_string += node.val
        return new_string
